### PR TITLE
update registry updates func

### PR DESCRIPTION
**Pull request by shayzluf**
[Part of] #2307 

---

# update registry updates func

**Write why you are making the changes in this pull request**
to sync with the new spec

**Write a summary of the changes you are making**
updated func:
UpdateRegistry

**Link anything that would be helpful or relevant to the reviewers**
https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#registry-updates

### CODE REVIEW

File: beacon-chain/core/validators/validator.go [hunk 1]
```diff
@@ -19,6 +20,11 @@ import (
 
 var log = logrus.WithField("prefix", "validator")
 
+type queueElement struct {
+	idx int
+	aEE uint64
```
Comment (terencechain) [inline beacon-chain/core/validators/validator.go]: what is queueElement and aEE, can you be more explicit with variable names?

File: beacon-chain/core/validators/validator.go [hunk 2]
```diff
@@ -310,105 +316,70 @@ func SlashValidator(state *pb.BeaconState, idx uint64) (*pb.BeaconState, error)
...
-			}
-			state, err = ActivateValidator(state, uint64(idx), false)
-			if err != nil {
-				return nil, fmt.Errorf("could not activate validator %d: %v", idx, err)
-			}
-			vStore.activatedValidators[updatedEpoch] =
-				append(vStore.activatedValidators[updatedEpoch], uint64(idx))
+		if validator.ActivationEligibilityEpoch == params.BeaconConfig().FarFutureEpoch &&
+			validator.EffectiveBalance >= params.BeaconConfig().MaxEffectiveBalance {
+			validator.ActivationEligibilityEpoch = currentEpoch
```
Comment (terencechain) [inline beacon-chain/core/validators/validator.go]: This code path is not tested, can we add test to cover this line?

File: beacon-chain/core/validators/validator.go [hunk 3]
```diff
@@ -310,105 +316,70 @@ func SlashValidator(state *pb.BeaconState, idx uint64) (*pb.BeaconState, error)
...
+			activationQueue = append(activationQueue, qe)
 		}
 	}
-	state.ValidatorRegistryUpdateEpoch = currentEpoch
-	return state, nil
+	sort.Slice(activationQueue, func(i, j int) bool { return activationQueue[i].aEE < activationQueue[j].aEE })
+	// dequeued validators for activation up to churn limit (without resetting activation epoch)
+	limit := uint64(len(activationQueue))
+	if helpers.ChurnLimit(state) < limit {
+		limit = helpers.ChurnLimit(state)
```
Comment (terencechain) [inline beacon-chain/core/validators/validator.go]: This line is not tested, can we add test to cover this line?

File: beacon-chain/core/validators/validator.go [hunk 4]
```diff
@@ -310,105 +316,70 @@ func SlashValidator(state *pb.BeaconState, idx uint64) (*pb.BeaconState, error)
...
+//         index for index, validator in enumerate(state.validator_registry) if
+//         validator.activation_eligibility_epoch != FAR_FUTURE_EPOCH and
+//         validator.activation_epoch >= get_delayed_activation_exit_epoch(state.finalized_epoch)
+//     ], key=lambda index: state.validator_registry[index].activation_eligibility_epoch)
+//     # Dequeued validators for activation up to churn limit (without resetting activation epoch)
+//     for index in activation_queue[:get_churn_limit(state)]:
+//         validator = state.validator_registry[index]
+//         if validator.activation_epoch == FAR_FUTURE_EPOCH:
+//             validator.activation_epoch = get_delayed_activation_exit_epoch(get_current_epoch(state))
+func UpdateRegistry(state *pb.BeaconState) *pb.BeaconState {
```
Comment (terencechain) [inline beacon-chain/core/validators/validator.go]: On a second thought, can we move this under `epoch_processing.go` and rename the function to `ProcessRegistryUpdates`? Thanks,

Comment (shayzluf) [reply beacon-chain/core/validators/validator.go]: done

File: beacon-chain/core/validators/validator.go (hunk 1 above)
Comment (shayzluf) [reply beacon-chain/core/validators/validator.go]: changed to ActivationEligibilityEpoch

File: beacon-chain/core/validators/validator.go (hunk 3 above)
Comment (shayzluf) [reply beacon-chain/core/validators/validator.go]: added tests

File: beacon-chain/core/validators/validator.go (hunk 2 above)
Comment (shayzluf) [reply beacon-chain/core/validators/validator.go]: adde tests, thanks

**Pull request merged**